packages/protocol/contracts/layer1/based/TaikoInbox.sol

```diff
@@ -90,94 +89,89 @@ abstract contract TaikoInbox is EssentialContract, ITaikoInbox, ITaiko {
+        unchecked {
+            meta_ = BatchMetadata({
+                txListHash: calldataUsed ? keccak256(_txList) : _calcTxListHash(_batchParams.numBlobs),
+                extraData: bytes32(uint256(config.baseFeeConfig.sharingPctg)),
+                coinbase: _coinbase,
+                batchId: stats2.numBatches,
+                gasLimit: config.blockMaxGasLimit,
+                timestamp: updatedParams.timestamp,
+                parentMetaHash: parentBatch.metaHash,
+                proposer: _proposer,
+                livenessBond: config.livenessBond,
+                proposedAt: uint64(block.timestamp),
+                proposedIn: uint64(block.number),
+                txListOffset: _batchParams.txListOffset,
+                txListSize: _batchParams.txListSize,
+                numBlobs: calldataUsed ? 0 : _batchParams.numBlobs,
+                anchorBlockId: updatedParams.anchorBlockId,
+                anchorBlockHash: blockhash(updatedParams.anchorBlockId),
+                signalSlots: _batchParams.signalSlots,
+                blocks: abi.encode(_batchParams.blocks),
+                anchorInput: _batchParams.anchorInput,
+                baseFeeConfig: config.baseFeeConfig
+            });
+        }
 
-            BlockV3 storage blk = state.blocks[stats2.numBlocks % config.blockRingBufferSize];
-            // SSTORE #1
-            blk.metaHash = metaHash;
+        require(meta_.txListHash != 0, BlobNotFound());
+        bytes32 metaHash = keccak256(abi.encode(meta_));
 
-            // SSTORE #2 {{
-            blk.blockId = stats2.numBlocks;
-            blk.timestamp = updatedParams.timestamp;
-            blk.anchorBlockId = updatedParams.anchorBlockId;
-            blk.nextTransitionId = 1;
-            blk.verifiedTransitionId = 0;
-            // SSTORE }}
+        Batch storage batch = state.batches[stats2.numBatches % config.batchRingBufferSize];
+        // SSTORE #1
+        batch.metaHash = metaHash;
 
-            // Update lastBlock to reference the most recently proposed block.
-            lastBlock = BlockInfo(metaHash, updatedParams.timestamp, updatedParams.anchorBlockId);
+        // SSTORE #2 {{
+        batch.batchId = stats2.numBatches;
+        batch.timestamp = updatedParams.timestamp;
+        batch.anchorBlockId = updatedParams.anchorBlockId;
+        batch.nextTransitionId = 1;
+        batch.numSubBlocks = uint8(_batchParams.blocks.length);
+        batch.verifiedTransitionId = 0;
+        // SSTORE }}
 
-            unchecked {
-                stats2.numBlocks += 1;
-                stats2.lastProposedIn = uint56(block.number);
-            }
-        } // end of for-loop
+        unchecked {
+            stats2.numBatches += 1;
+            stats2.lastProposedIn = uint56(block.number);
+        }
 
-        _debitBond(_proposer, config.livenessBond * _paramsArray.length);
-        emit BlocksProposedV3(metas_, calldataUsed, _txList);
+        _debitBond(_proposer, config.livenessBond);
```

Comment:
So rn for each batch, we only charge `config.livenessBond` once, is it by design?

Reply:
Yes, as a batch is the smallest unit to be proved, so we only debit liveness bond once per batch.